frame/staking/src/lib.rs

```diff
@@ -2063,50 +2081,58 @@ decl_module! {
 		///    minimized (to ensure less variance)
 		///
 		/// # <weight>
-		/// E: number of edges. m: size of winner committee. n: number of nominators. d: edge degree
-		/// (16 for now) v: number of on-chain validator candidates.
-		///
-		/// NOTE: given a solution which is reduced, we can enable a new check the ensure `|E| < n +
-		/// m`. We don't do this _yet_, but our offchain worker code executes it nonetheless.
-		///
-		/// major steps (all done in `check_and_replace_solution`):
-		///
-		/// - Storage: O(1) read `ElectionStatus`.
-		/// - Storage: O(1) read `PhragmenScore`.
-		/// - Storage: O(1) read `ValidatorCount`.
-		/// - Storage: O(1) length read from `SnapshotValidators`.
-		///
-		/// - Storage: O(v) reads of `AccountId` to fetch `snapshot_validators`.
-		/// - Memory: O(m) iterations to map winner index to validator id.
-		/// - Storage: O(n) reads `AccountId` to fetch `snapshot_nominators`.
-		/// - Memory: O(n + m) reads to map index to `AccountId` for un-compact.
-		///
-		/// - Storage: O(e) accountid reads from `Nomination` to read correct nominations.
-		/// - Storage: O(e) calls into `slashable_balance_of_vote_weight` to convert ratio to staked.
-		///
-		/// - Memory: build_support_map. O(e).
-		/// - Memory: evaluate_support: O(E).
-		///
-		/// - Storage: O(e) writes to `QueuedElected`.
-		/// - Storage: O(1) write to `QueuedScore`
-		///
-		/// The weight of this call is 1/10th of the blocks total weight.
+		/// All weight notes are pertaining to the case of a better solution, in which we execute
+		/// the longest code path.
```

Comment:
> if nominations are nominating validator that are not part of the winner,

This will be caught mid-flight as the part of the validatoion and the execution stops. 

Unless if someone submits a solution which is maliciously big, but invalid. For that, we're safe. The weight formula is a function of both the size of the snapshot, and the size of the solution itself (see how I use and implement `compact.edge_count()` and `compact.len()`). So you can only ever submit a big solution, until the weight limit, in which case you still pay for it, so we're good.

Reply:
> This will be caught mid-flight as the part of the validatoion and the execution stops.

I only mean that if one submit a solution with:
* 1 nominator which nominates 16 different validator
* 1 validator which nominate himself.
* only one winner

then in the following loop https://github.com/paritytech/substrate/blob/5a7600912b7fd50091990899e8ad93ad618f40c4/frame/staking/src/lib.rs#L2608-L2656
We read 16 slashing span (maybe I missed an earlier check inside the code but AFAIK this what should happen no ?).

But if we do:

> All weight notes are pertaining to the case of a better solution, in which we execute the longest code path.

Then if we weight for the better solution, that means this is a valid solution, thus the number of slashing spans is at most the number of winners (in our example 1)

Also:

currently it is weighted with:
```rust
			.saturating_add( // SlashingSpans
				T::DbWeight::get().reads(
					compact.len()
						.saturating_sub(winners.len())
						.saturating_mul(compact.average_edge_count())
						as Weight
				)
			)
```
So it will pay only for reading: `(2 - 1)*(17/2)=8` instead of 16.


In general the weight should pay for the worst case for the given input.

Reply:
> I only mean that if one submit a solution with:
> 1 nominator which nominates 16 different validator
> 1 validator which nominate himself.
> only one winner

This is not a valid solution. It is not really clear, but the function `build_support_map` returns an error if the `Assignment` contains validators are not part of the `winners`. In your example, since we want only one elected winner, the assignment with 16 validators is invalid and will be caught. 

> So it will pay only for reading: (2 - 1)*(17/2)=8 instead of 16.

This I agree with. Initially I was always assuming that each assignment has 16 votes. Now I take an average. I can either

1- go back to taking 16. 
2- write a new method that actually gives the exact number of times we read the slashing spans. Basically I can iterate over the compact, and sum all the edge counts for those that are not a self-vote. We _could_ also assume that we always have exactly `winners.len()` self vote but that could be a later optimisation.

Reply:
> In your example, since we want only one elected winner, the assignment with 16 validators is invalid and will be caught.

I know, my point is that it will be caught after having read 16 SlashingSpan, so those 16 reads must be paid for.

Reply:
wait, I think we're totally wrong here. 

SlashingSpan is keyed by `Validators`. Duplicate reads should not be counted, right? Then we always read exactly `winners.len()` keys from `SlashingSpans`.

Reply:
> [..] We could also assume that we always have exactly winners.len() self vote but that could be a later optimisation.

Only in a valid solution

Reply:
> SlashingSpan is keyed by Validators. Duplicate reads should not be counted, right? Then we always read exactly winners.len() keys from SlashingSpans.

true in a valid solution, but false in an invalid solution, and the checks happens after having read the SlashinSpans

Reply:
Okay what I will do without changing the order of checks is: We will charge for the exact number of edges. So if you submit anything invalid you will still pay for it. And we accumulate refund per self vote that we process.